### ISSUE TITLE
Make hbal still improve situation when exceeding oversubscription ratios

### ISSUE THREAD

**Issue by Ganeti-Issues-Migrator** [ganeti/ganeti#750]
Originally reported of Google Code with ID 694.

```
Reported by twa in #ganeti

<b>What steps will reproduce the problem?</b>

Consider the attached cluster configuration;
by means of migration, all instances are running
on the nodes different from dv-04. Ask hbal to balance
it.

<b>What is the expected output? What do you see instead?</b>

I'd expect some migration (f) moves so instances
are running on node dv-04. Instead hbal does not
find any solution.

<b>Please use labels and text to provide additional information.</b>

```
Originally added on 2014-01-28 09:09:21 +0000 UTC.

Attachments: 
[cluster.text](https://storage.googleapis.com/google-code-attachments/ganeti/issue-694/comment-0/cluster.text)

**Comment by Ganeti-Issues-Migrator** [comment]
```
The cluster is already in a policy-violating state: the oversubscription rate on spindles is exceeded. Every move will leave the cluster in that invalid state, so is not considered possible. Changing the policy by increasing the oversubscription rate on splindles allows finding a solution again.
```
Originally added on 2014-01-29 12:12:35 +0000 UTC.

**Comment by Ganeti-Issues-Migrator** [comment]
```
Rewording as feature request:

 When a cluster is in a fully functioning but policy-wise
 bad state (like exceeding oversubscription ratios on vCPUs
 or spindles), it would be desirable that hbal still carries
 out those operations that
 - do not make the policy violation any worse
 - improve the cluster status

Additionally, the diagnostic messages of hbal in that situation
could be improved.
```
Originally added on 2014-01-29 12:18:13 +0000 UTC.

Added Labels: Type-Enhancement
Removed Labels: Type-Defect
Added to Milestone: Release2.12
Removed from Milestone: Release2.9

**Comment by Ganeti-Issues-Migrator** [comment]
```
feaure freeze for 2.12 over, postponing
```
Originally added on 2014-04-24 14:21:33 +0000 UTC.

Added to Milestone: Release2.13
Removed from Milestone: Release2.12

**Comment by Ganeti-Issues-Migrator** [comment]
```
commit 9ebd09a2e26872e86c0892e6da6494a6410b457d
Author: Klaus Aehlig <aehlig@google.com>
Date:   Tue Jul 1 15:14:25 2014 +0200

    Document --ignore-soft-errors in man pages
    
    Describe that this options lifts restrictions on the considered
    moves and thus allow progress in an all-bad situation.
    
    Signed-off-by: Klaus Aehlig <aehlig@google.com>
    Reviewed-by: Petr Pudlak <pudlak@google.com>

commit 02b1e840b1d0bd99f207a4d91509b8eef73fc5cf
Author: Klaus Aehlig <aehlig@google.com>
Date:   Tue Jul 1 14:33:58 2014 +0200

    Add a test for hbal under soft errors
    
    The example was submitted as part of the report for issue 649.
    It shows a cluster, where every node exceeds the oversubscription
    ratio on spindles and, on top of that, is heavily inbalanced.
    While the oversubscription violation cannot be removed, the
    balancedness can be improved significantly.
    
    Signed-off-by: Klaus Aehlig <aehlig@google.com>
    Reviewed-by: Petr Pudlak <pudlak@google.com>

commit 80f0c5299a5d17c2c79dcd361c31fa421e26820f
Author: Klaus Aehlig <aehlig@google.com>
Date:   Tue Jul 1 13:48:46 2014 +0200

    Make htools ignore soft errors if desired
    
    Make htools support the --ignore-soft-errors options,
    telling to not care about soft errors during balancing.
    This might be necessary if the cluster is so heavily
    overloaded on oversubscribable resources that otherwise
    no movement would be possible. Note that the balancing
    metric is still designed to minimize the amount of
    soft errors.
    
    Signed-off-by: Klaus Aehlig <aehlig@google.com>
    Reviewed-by: Petr Pudlak <pudlak@google.com>

commit b4aa31ec640f098bc85f3fa06cafbb8139083b45
Author: Klaus Aehlig <aehlig@google.com>
Date:   Tue Jul 1 13:41:06 2014 +0200

    Add an option --ignore-soft-errors
    
    ...that will tell htools to no care about any soft
    errors (like CPU oversubscription ratio) in balancing
    attempts.
    
    Signed-off-by: Klaus Aehlig <aehlig@google.com>
    Reviewed-by: Petr Pudlak <pudlak@google.com>

commit 4f4c4d7c52619d1fd20f7d1c76106d5b144384f7
Author: Klaus Aehlig <aehlig@google.com>
Date:   Tue Jul 1 10:21:18 2014 +0200

    Extend tryBalance with an option to ignore soft errors
    
    Generally, htools never considers moves that would yield a soft
    error (exceeding oversubscription ratios, N+1 violatsion) unless
    the instance moved has its primary on an offline mode. While this
    is useful in general to keep the cluster free of soft errors, it
    sometimes leaves to little room for improvement in a situation where
    soft errors are unavoidable. Hence add a flag to unconditionally
    ignore soft errors.
    
    Signed-off-by: Klaus Aehlig <aehlig@google.com>
    Reviewed-by: Petr Pudlak <pudlak@google.com>

commit 52fbac619c27497e43b188a33ccb4bce58ac14c8
Author: Klaus Aehlig <aehlig@google.com>
Date:   Mon Jun 30 17:00:07 2014 +0200

    Warn about nodes policy wise in a bad state
    
    Add a warning on startup of hbal, if nodes are in a policy-wise
    bad state. Hbal will never do any moves where the accepting
    node afterwards is in a policy-violating state. A consequence of
    this is that hbal will not improve the situation if a cluster has
    too many policy-wise bad states (issue 694). So make the user aware
    of this situation.
    
    Signed-off-by: Klaus Aehlig <aehlig@google.com>
    Reviewed-by: Petr Pudlak <pudlak@google.com>

commit 2618f70c2cd2a434c9ccfb782acb38759e08cbff
Author: Klaus Aehlig <aehlig@google.com>
Date:   Mon Jun 30 14:12:50 2014 +0200

    Add a function testing policy health of a node
    
    Add a function that tests whether a node is in a healthy state
    policy wise.  The type is chosen to reflect that of a no-op
    operation on a node.
    
    Signed-off-by: Klaus Aehlig <aehlig@google.com>
    Reviewed-by: Petr Pudlak <pudlak@google.com>

```
Originally added on 2014-07-02 10:28:33 +0000 UTC.

Mentioned Issue with GoogleCode ID 649 has been migrated to GitHub with ID #705.
Mentioned Issue with GoogleCode ID 694 has been migrated to GitHub with ID #750.
Changed State: Fixed

**Comment by Ganeti-Issues-Migrator** [comment]
```
Released as part of 2.14.0
```
Originally added on 2015-06-02 10:37:02 +0000 UTC.

Changed State: Released
